platforms/software/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/DefaultCompositeTestEventGenerator.java

```diff
@@ -0,0 +1,137 @@
+    private final TestResultProcessor processor;
+    private final IdGenerator<?> idGenerator;
+    private final @Nullable TestDescriptorInternal parent;
+    private final TestDescriptorInternal testDescriptor;
+    private Set<DefaultCompositeTestEventGenerator> children;
+    private boolean closed;
+
+    public DefaultCompositeTestEventGenerator(
+        TestResultProcessor processor, IdGenerator<?> idGenerator, @Nullable TestDescriptorInternal parent, TestDescriptorInternal testDescriptor
+    ) {
+        this.processor = processor;
+        this.idGenerator = idGenerator;
+        this.parent = parent;
+        this.testDescriptor = testDescriptor;
+    }
+
+    private void requireOpen() {
+        if (closed) {
+            throw new IllegalStateException("Test event generator is closed");
+        }
+    }
+
+    private void addChild(DefaultCompositeTestEventGenerator child) {
+        if (children == null) {
+            children = Collections.newSetFromMap(new WeakHashMap<>());
+        }
+        children.add(child);
+    }
+
+    @Override
+    public TestEventGenerator createAtomicNode(String name, String displayName) {
+        requireOpen();
+        DefaultCompositeTestEventGenerator child = new DefaultCompositeTestEventGenerator(
+            processor, idGenerator, testDescriptor, new DefaultTestDescriptor(idGenerator.generateId(), null, name, null, displayName)
+        );
+        addChild(child);
+        return child;
+    }
+
+    @Override
+    public CompositeTestEventGenerator createCompositeNode(String name) {
+        requireOpen();
+        DefaultCompositeTestEventGenerator child = new DefaultCompositeTestEventGenerator(
+            processor, idGenerator, testDescriptor, new DefaultTestSuiteDescriptor(idGenerator.generateId(), name)
+        );
+        addChild(child);
+        return child;
+    }
+
+    @Override
+    public void started(Instant startTime) {
+        requireOpen();
+        processor.started(testDescriptor, new TestStartEvent(startTime.toEpochMilli(), parent == null ? null : parent.getId()));
+    }
+
+    @Override
+    public void output(TestOutputEvent.Destination destination, String output) {
+        requireOpen();
+        processor.output(testDescriptor.getId(), new DefaultTestOutputEvent(destination, output));
+    }
+
+    @Override
+    public void failure(Throwable failure) {
+        requireOpen();
+        processor.failure(testDescriptor.getId(), TestFailure.fromTestFrameworkFailure(failure));
+    }
+
+    @Override
+    public void completed(Instant endTime, TestResult.ResultType resultType) {
+        requireOpen();
+        processor.completed(testDescriptor.getId(), new TestCompleteEvent(endTime.toEpochMilli(), resultType));
+    }
+
+    @Override
+    public void close() {
+        if (closed) {
+            return;
+        }
+        closed = true;
+        if (children != null) {
+            List<Throwable> errors = null;
```

Comment:
Instead of tracking errors like this, can the root most TestEventService have a generic listener that keeps track of any failure and go off of that?

Alternatively, we could pass a "listener" into each child that we make sure is called when there's a failure. that listener could be implemented by the root most TestEventGenerator.

Reply:
As per my question in the spec, I'm not sure we want to fail on any failure. I've left it to the root node for now.